File: src/data/courses/dsa/prefix-sum/prefix-sum-fundamentals/longest-subarray-sum-equals-k/code/java/LongestSubarraySumEqualsK.java
import java.util.*;

public class LongestSubarraySumEqualsK {
    
    // ==================== SOLUTION 1: BRUTE FORCE ====================
    // Time: O(N³) | Space: O(1)
    
    public static int longestSubarrayBrute(int[] nums, int k) {
        int n = nums.length;
        int maxLen = 0;
        
        for (int i = 0; i < n; i++) {
            for (int j = i; j < n; j++) {
                int subarraySum = 0;
                for (int idx = i; idx <= j; idx++) {
                    subarraySum += nums[idx];
                }
                if (subarraySum == k) {
                    maxLen = Math.max(maxLen, j - i + 1);
                }
            }
        }
        
        return maxLen;
    }
    
    
    // ==================== SOLUTION 2: BRUTE FORCE OPTIMIZED ====================
    // Time: O(N²) | Space: O(1)
    
    public static int longestSubarrayOptimized(int[] nums, int k) {
        int n = nums.length;
        int maxLen = 0;
        
        for (int i = 0; i < n; i++) {
            int currentSum = 0;
            for (int j = i; j < n; j++) {
                currentSum += nums[j];
                if (currentSum == k) {
                    maxLen = Math.max(maxLen, j - i + 1);
                }
            }
        }
        
        return maxLen;
    }
    
    
    // ==================== SOLUTION 3: PREFIX SUM WITH HASHMAP ====================
    // Time: O(N) | Space: O(N)
    
    public static int longestSubarrayPrefixSum(int[] nums, int k) {
        Map<Integer, Integer> prefixSumMap = new HashMap<>();
        prefixSumMap.put(0, -1);  // Empty prefix before index 0
        
        int prefixSum = 0;
        int maxLen = 0;
        
        for (int i = 0; i < nums.length; i++) {
            prefixSum += nums[i];
            
            // Check if (prefixSum - k) exists
            if (prefixSumMap.containsKey(prefixSum - k)) {
                int length = i - prefixSumMap.get(prefixSum - k);
                maxLen = Math.max(maxLen, length);
            }
            
            // Store first occurrence only (to maximize length)
            if (!prefixSumMap.containsKey(prefixSum)) {
                prefixSumMap.put(prefixSum, i);
            }
        }
        
        return maxLen;
    }
    
    
    // ==================== SOLUTION 4: SLIDING WINDOW (POSITIVE ONLY) ====================
    // Time: O(N) | Space: O(1)
    
    public static int longestSubarraySlidingWindow(int[] nums, int k) {
        int left = 0;
        int currentSum = 0;
        int maxLen = 0;
        
        for (int right = 0; right < nums.length; right++) {
            currentSum += nums[right];
            
            while (currentSum > k && left <= right) {
                currentSum -= nums[left];
                left++;
            }
            
            if (currentSum == k) {
                maxLen = Math.max(maxLen, right - left + 1);
            }
        }
        
        return maxLen;
    }
    
    
    // ==================== MAIN SOLUTION (RECOMMENDED) ====================
    
    public static int longestSubarraySumK(int[] nums, int k) {
        /*
         * Find length of longest subarray with sum equal to k.
         * 
         * Time: O(N) - single pass
         * Space: O(N) - HashMap
         * 
         * Works with negative numbers and zeros.
         */
        return longestSubarrayPrefixSum(nums, k);
    }
}

File: src/data/courses/dsa/prefix-sum/prefix-sum-fundamentals/longest-subarray-sum-equals-k/code/python/longest_subarray_sum_equals_k.py
# ==================== SOLUTION 1: BRUTE FORCE ====================
# Time: O(N³) | Space: O(1)

def longest_subarray_brute(nums, k):
    n = len(nums)
    max_len = 0
    
    for i in range(n):
        for j in range(i, n):
            subarray_sum = sum(nums[i:j+1])
            if subarray_sum == k:
                max_len = max(max_len, j - i + 1)
    
    return max_len


# ==================== SOLUTION 2: BRUTE FORCE OPTIMIZED ====================
# Time: O(N²) | Space: O(1)

def longest_subarray_optimized(nums, k):
    n = len(nums)
    max_len = 0
    
    for i in range(n):
        current_sum = 0
        for j in range(i, n):
            current_sum += nums[j]
            if current_sum == k:
                max_len = max(max_len, j - i + 1)
    
    return max_len


# ==================== SOLUTION 3: PREFIX SUM WITH HASHMAP ====================
# Time: O(N) | Space: O(N)

def longest_subarray_prefix_sum(nums, k):
    prefix_sum_map = {0: -1}  # Empty prefix before index 0
    prefix_sum = 0
    max_len = 0
    
    for i in range(len(nums)):
        prefix_sum += nums[i]
        
        # Check if (prefixSum - k) exists
        if (prefix_sum - k) in prefix_sum_map:
            length = i - prefix_sum_map[prefix_sum - k]
            max_len = max(max_len, length)
        
        # Store first occurrence only (to maximize length)
        if prefix_sum not in prefix_sum_map:
            prefix_sum_map[prefix_sum] = i
    
    return max_len


# ==================== SOLUTION 4: SLIDING WINDOW (POSITIVE ONLY) ====================
# Time: O(N) | Space: O(1)

def longest_subarray_sliding_window(nums, k):
    left = 0
    current_sum = 0
    max_len = 0
    
    for right in range(len(nums)):
        current_sum += nums[right]
        
        while current_sum > k and left <= right:
            current_sum -= nums[left]
            left += 1
        
        if current_sum == k:
            max_len = max(max_len, right - left + 1)
    
    return max_len


# ==================== MAIN SOLUTION (RECOMMENDED) ====================

def longestSubarraySumK(nums, k):
    """
    Find length of longest subarray with sum equal to k.
    
    Time: O(N) - single pass
    Space: O(N) - HashMap
    
    Works with negative numbers and zeros.
    """
    return longest_subarray_prefix_sum(nums, k)
